server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java

```diff
@@ -0,0 +1,202 @@
+        Field pathField = new Field(TSDataType.TEXT);
+        pathField.setBinaryV(new Binary(selectedSeries.get(i).getFullPath()));
+        resultRecord.addField(pathField);
+
+        Field valueField = new Field(TSDataType.TEXT);
+        valueField.setBinaryV(new Binary(lastTimeValuePair.getValue().getStringValue()));
+        resultRecord.addField(valueField);
+
+        dataSet.putRecord(resultRecord);
+      }
+    }
+
+    return dataSet;
+  }
+
+  /**
+   * get last result for one series
+   *
+   * @param context query context
+   * @return TimeValuePair
+   */
+  private TimeValuePair calculateLastPairForOneSeries(
+      Path seriesPath, TSDataType tsDataType, QueryContext context)
+      throws IOException, QueryProcessException, StorageEngineException {
+
+    // Retrieve last value from MNode
+    MNode node = null;
+    try {
+      node = MManager.getInstance().getDeviceNodeWithAutoCreateStorageGroup(seriesPath.toString());
+    } catch (MetadataException e) {
+      throw new QueryProcessException(e);
+    }
+    if (((LeafMNode)node).getCachedLast() != null) {
+      return ((LeafMNode)node).getCachedLast();
+    }
+
+    QueryDataSource dataSource =
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, null);
+
+    List<TsFileResource> seqFileResources = dataSource.getSeqResources();
+    List<TsFileResource> unseqFileResources = dataSource.getUnseqResources();
+
+    TimeValuePair resultPair = new TimeValuePair(Long.MIN_VALUE, null);
+
+    for (int i = seqFileResources.size() - 1; i >= 0; i--) {
+      List<ChunkMetaData> chunkMetadata =
+          loadSatisfiedChunkMetadata(seqFileResources.get(i), seriesPath, context);
+      if (!chunkMetadata.isEmpty()) {
+        ChunkMetaData lastChunkMetaData = chunkMetadata.get(chunkMetadata.size() - 1);
+        Statistics chunkStatistics = lastChunkMetaData.getStatistics();
+        resultPair = constructLastPair(
+                chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), tsDataType);
+        break;
+      }
+    }
+
+    long version = 0;
+    for (TsFileResource resource : unseqFileResources) {
+      if (resource.getEndTimeMap().get(seriesPath.getDevice()) < resultPair.getTimestamp()) {
+        break;
+      }
+      List<ChunkMetaData> chunkMetadata = loadSatisfiedChunkMetadata(resource, seriesPath, context);
+      for (ChunkMetaData chunkMetaData : chunkMetadata) {
+        if (chunkMetaData.getEndTime() == resultPair.getTimestamp()
+            && chunkMetaData.getVersion() > version) {
+          Statistics chunkStatistics = chunkMetaData.getStatistics();
+          resultPair = constructLastPair(
+                  chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), tsDataType);
+          version = chunkMetaData.getVersion();
+        }
+      }
+    }
+
+    // Update cached last value with low priority
+    ((LeafMNode)node).updateCachedLast(resultPair, false, Long.MIN_VALUE);
```

Comment:
when I saw this line, I think maybe move the lastFlushTime out of the updateCachedLast mehod and check in the insertion is better